include/oneapi/dpl/pstl/parallel_backend_utils.h

```diff
@@ -301,6 +306,72 @@ __set_symmetric_difference_construct(_ForwardIterator1 __first1, _ForwardIterato
     return __cc_range(__first2, __last2, __result);
 }
 
+template <typename _ValueType, typename _GetNumThreads, typename _GetThreadNum, typename... _Args>
+struct __enumerable_thread_local_storage
+{
+    template <typename... _LocalArgs>
+    __enumerable_thread_local_storage(_LocalArgs&&... __args)
+        : __thread_specific_storage(_GetNumThreads{}()), __num_elements(0), __args(std::forward<_LocalArgs>(__args)...)
+    {
+    }
+
+    // Note: size should not be used concurrently with parallel loops which may instantiate storage objects, as it may
+    // not return an accurate count of instantiated storage objects in lockstep with the number allocated and stored.
+    // This is because the count is not atomic with the allocation and storage of the storage objects.
+    std::size_t
+    size() const
+    {
+        // only count storage which has been instantiated
+        return __num_elements.load();
```

Comment:
I am not sure but probably we have small reserve here: may be the default param value [std::memory_order_seq_cst](http://en.cppreference.com/w/cpp/atomic/memory_order) may be changed.

Reply:
I suppose here we could use `memory_order_relaxed`.  Do you agree?

Reply:
I think so. @akukanov how do you think?

Reply:
Given that we "require" `size()` to only be used in serial parts of code, and so there is no need to establish happens-before relations of particular expressions, I believe all operations with the atomic counter can be `relaxed`, including increments.